rust/agama-server/src/agama-web-server.rs

```diff
@@ -32,23 +60,213 @@ struct Cli {
+        tls_builder
+            .set_certificate_file(PathBuf::from(cert_file), SslFiletype::PEM)
+            .unwrap();
+
+        tracing::info!("Loading PEM key: {}", key_file);
+        tls_builder
+            .set_private_key_file(PathBuf::from(key_file), SslFiletype::PEM)
+            .unwrap();
+    }
+
+    // check that the key belongs to the certificate
+    tls_builder.check_private_key().unwrap();
+
+    tls_builder.build()
+}
+
+// build a response for the HTTP -> HTTPS redirection
+// returns 308 permanent redirect
+fn redirect_https(host: &str, uri: &hyper::Uri) -> Response<String> {
+    let builder = Response::builder()
+        // build the redirection target URL
+        .header("Location", format!("https://{}{}", host, uri))
+        .status(hyper::StatusCode::PERMANENT_REDIRECT);
+
+    builder.body(String::from("")).unwrap()
+}
+
+// build an error response for the HTTP -> HTTPS redirection when we cannot build
+// the redirect response from the original request
+// returns error 400
+fn redirect_error() -> Response<String> {
+    let builder = Response::builder().status(hyper::StatusCode::BAD_REQUEST);
+
+    let msg = "HTTP protocol is not allowed for external requests, please use HTTPS.\n";
+    builder.body(String::from(msg)).unwrap()
+}
+
+// build a router for the HTTP -> HTTPS redirection
+// if the redirection URL cannot be built from the original request it returns error 400
+// instead of the redirection
+fn https_redirect() -> Router {
+    // see https://docs.rs/axum/latest/axum/routing/struct.Router.html#example
+    let redirect_service = tower::service_fn(|req: axum::extract::Request| async move {
+        let request_host = req.headers().get("host");
+        match request_host {
+            // missing host in the request header, we cannot build the redirection URL, return error 400
+            None => Ok(redirect_error()),
+            Some(host) => {
+                let host_string = host.to_str();
+                match host_string {
+                    // invalid host value in the request
+                    Err(_) => Ok(redirect_error()),
+                    Ok(host_str) => return Ok(redirect_https(host_str, req.uri())),
+                }
+            }
+        }
+    });
+
+    Router::new()
+        // the wildcard path below does not match an empty path, we need to match it explicitly
+        .route_service("/", redirect_service)
+        .route_service("/*path", redirect_service)
+}
+
+// start the web server
+async fn start_server(address: String, service: Router, ssl_acceptor: SslAcceptor) {
+    tracing::info!("Starting Agama web server at {}", address);
+
+    // see https://github.com/tokio-rs/axum/blob/main/examples/low-level-openssl/src/main.rs
+    // how to use axum with openSSL
+    let listener = tokio::net::TcpListener::bind(&address)
+        .await
+        .unwrap_or_else(|_| panic!("could not listen on {}", &address));
```

Comment:
will it also print details why it cannot listed like port occupied or insuficient permissions?

Reply:
just a question? Why do you consider ```unwrap_or_else``` better then ```expect``` I put there in previous commit? I know that documentation states that "Because this function may panic, its use is generally discouraged." But as you use ```panic!``` macro anyway you more less mimic expect here.